### ISSUE TITLE
Memory StateManager should have some sort of expiration

### ISSUE THREAD

**Issue by picklelo** [reflex-dev/reflex#3021]
The state manager is essentially a dictionary mapping between a user token (browser tab) and their state.

When using Redis as a state manager (by setting the `redis_url` in the `rxconfig.py`) there is a configurable expiration time (by default 30 mins). However when using the in memory state manager, states will accumulate until eventually there is an OOM error. 

We should implement some sort of expiration or LRU caching on the memory state manager to make this usable in production.

**Comment by AmanSal1** [comment]
@picklelo So, I've been considering giving it a try. What I've figured out so far is that we need to modify the **get** and **set** methods, possibly adding additional methods to the **StateManagerMemory** class, and then write test cases for it. In my opinion, implementing **LRU caching** would be a more complex approach. If the app works fine, **expiration-based caching** would be a great approach, as we only have to maintain the time period associated with each state and accordingly expire and remove it from memory. With your guidance, I think I can accomplish this

**Comment by picklelo** [comment]
@AmanSal1 thanks for looking into this! Agree with you that expiration-based caching would be ideal, but I'm unsure how to implement that easily, do you have a way?

**Comment by picklelo** [comment]
> Periodically, we'll clean up expired states using a method that iterates through the expiration times dictionary and removes expired states.

This is the part that seems a bit tricky to me, there would have to be some worker thread that keeps checking and deleting this. I wonder if for a first pass the LRU cache can be a good simple approach that achieves most of the goal, and we can expand to the expiration method in a follow up. You mentioned LRU may be more complicated - why do you think so? I feel it should be easier than this method.

**Comment by AmanSal1** [comment]
@picklelo  So like why I said that LRU caching is bit difficult because of the difference in eviction strategies contributes to the perception that implementing LRU caching may be more complex than expiration-based caching. In expiration-based caching, the eviction process is automatic and deterministic, driven solely by the expiration times of items. In contrast, in LRU caching, you need to actively manage the access order of items and make decisions about which items to evict based on access history. Additionally we also should consider the use cases of our app like in both the methods  have their own trade offs.

**Comment by picklelo** [comment]
It looks like there are available packages for both ways:

LRU Dict: https://pypi.org/project/lru-dict/
Expiring Dict: https://pypi.org/project/expiringdict/

So maybe we can go with one of these, either using the library or reimplementing some of the logic within reflex. But based on this, expiring dict makes the most sense as it will match our Redis API, and we can have a parameter for the expiration time.

**Comment by AmanSal1** [comment]
@picklelo Yes, I also researched this and found the LRU cache package from functools . But when you said that LRU caching could be simple, I think in some sense maybe you are correct. Because what I've gathered is that we just need a decorator over the get_state method, and everything else is handled by the LRU caching library. Maybe at the implementation level, it may differ, but yes, it could be simple. Can you tell me how we can test these StateManagerMemory instances? What is the right way to create their instance and test them?

**Comment by picklelo** [comment]
@AmanSal1 We have a test on the Redis expiration, we can extend something similar for the Memory one. See [here](https://github.com/reflex-dev/reflex/blob/d5e3ccaed87efa7e7bc0e658f9fe12ceb3f9a93f/tests/test_state.py#L1622)

**Comment by AmanSal1** [comment]
@picklelo Is this the right way to declare StatManagerMemory instance .
```

import pytest
from reflex.state import StateManagerMemory


@pytest.mark.asyncio
async def test_state_manager_memory_caching():
    """Test the LRU caching mechanism in StateManagerMemory."""

    # Create an instance of StateManagerMemory
    state_manager = StateManagerMemory()

    assert state_manager.get_state.cache_info().hits == 0
    assert state_manager.get_state.cache_info().misses == 0

    token = "test_token"
    for _ in range(5):
        state = await state_manager.get_state(token)

    # Ensure that all calls hit the cache after the first call
    assert state_manager.get_state.cache_info().hits == 4
    assert state_manager.get_state.cache_info().misses == 1

    # Call get_state with a different token
    for _ in range(5):
        state = await state_manager.get_state("another_token")

    # Ensure that new calls with a different token also hit the cache
    assert state_manager.get_state.cache_info().hits == 8
    assert state_manager.get_state.cache_info().misses == 2

```

**Comment by picklelo** [comment]
@AmanSal1 you need to pass in a State class to it:

```python
class ExampleState(rx.State):
   var1: int
   var2: int

state_manager = StateManagerMemory(state=ExampleState)
```

The other thing to check would be that when it receives it from a cache hit it retains the old state values, while when there's a cache miss it will get a fresh state.

**Comment by AmanSal1** [comment]
@picklelo So like I wanted to know what is the  **Can't instantiate abstract class StateManagerMemory with abstract method get_state** error because if I run pytest test_state.py I get multiple errors I am adding a screenshot of this ?

![2](https://github.com/reflex-dev/reflex/assets/121633121/139bfb5f-03e1-4d3f-aa3c-44716002c161)

**Comment by picklelo** [comment]
@AmanSal1 what Python version are you on? This is without modifying it at all? The tests should pass as we run CI for all python versions.

**Comment by AmanSal1** [comment]
@picklelo When I make a change in the test_state.py file and then revert it back and rerun pytest, the test fails. As seen in the photo above, something seems to be initiated or perhaps cached, causing the test to fail repeatedly.

**Comment by ElijahAhianyo** [comment]
> @picklelo So like I wanted to know what is the **Can't instantiate abstract class StateManagerMemory with abstract method get_state** error because if I run pytest test_state.py I get multiple errors I am adding a screenshot of this ?
> 
> ![2](https://private-user-images.githubusercontent.com/121633121/321289139-139bfb5f-03e1-4d3f-aa3c-44716002c161.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyNTgwMTgsIm5iZiI6MTcxMzI1NzcxOCwicGF0aCI6Ii8xMjE2MzMxMjEvMzIxMjg5MTM5LTEzOWJmYjVmLTAzZTEtNGQzZi1hYTNjLTQ0NzE2MDAyYzE2MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNlQwODU1MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZTljZGFhMThkYWI1YTZhZTExNWIxOWRmMmQwMjZlYzRiMzc4ZjdhOGIzOTYxNzZiMDg5Mzk0MTAxOTZmOGNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WxwOs9klaXq8gUSuq2Hd2wR5jgSbXF52nBdYAPm-EIw)

@AmanSal1 The tests should run as is without any modifications, The redis tests should also be skipped. Are you running the tests with redis?

**Comment by AmanSal1** [comment]
@ElijahAhianyo Yes, I understand your point. I'm attempting to integrate LRU caching into the built-in state manager. However, when I write the test case in test_state.py, it fails. Even if I revert my changes, the test continues to fail. Initially, if we don't change anything, it exhibits the same behavior you mentioned

**Comment by ElijahAhianyo** [comment]
> @ElijahAhianyo Yes, I understand your point. I'm attempting to integrate LRU caching into the built-in state manager. However, when I write the test case in test_state.py, it fails. Even if I revert my changes, the test continues to fail. Initially, if we don't change anything, it exhibits the same behavior you mentioned

I see, what's the code diff you're adding? Will be really helpful if there's a linked PR I can try to repro

**Comment by AmanSal1** [comment]
@ElijahAhianyo Sure . Will be adding a PR shortly .

**Comment by AmanSal1** [comment]
@ElijahAhianyo I have made a PR .

**Comment by ElijahAhianyo** [comment]
> @ElijahAhianyo I have made a PR .

It looks like the issue youre facing is pydantic related. There's a weird behavior in v1 (not an issue in v2)where it looks like the meta data of the function decorated with `lru_cache` is not copied properly and works if the abstract method is also decorated.
Eg:

```python
from pydantic import BaseModel
from ABC import ABC, abstractmethod
from functools import lru_cache

class AbstractClass(BaseModel, ABC):

    @abstractmethod
    def abstract_method(self, token):
        pass

    class Config:
        """Pydantic config."""
        arbitrary_types_allowed = True


class ChildClass(AbstractClass):

    @lru_cache(maxsize=100)
    def abstract_method(self, token):
        pass

child_class = ChildClass()
```
This above will throw an error. But below works:

```python
...
class AbstractClass(BaseModel, ABC):
   @lru_cache(maxsize=100)
    @abstractmethod
    def abstract_method(self, token):
        pass

    class Config:
        """Pydantic config."""
        arbitrary_types_allowed = True


class ChildClass(AbstractClass):

    @lru_cache(maxsize=100)
    def abstract_method(self, token):
        pass

child_class = ChildClass()
```

For the weird meta data behavior, here's a sample:

```python
# pydantic v1
...
class AbstractClass(BaseModel, ABC):

    @lru_cache(maxsize=100)
    @abstractmethod
    def abstract_method(self, token):
        pass

    class Config:
        """Pydantic config."""
        arbitrary_types_allowed = True


class ChildClass(AbstractClass):

    @lru_cache(maxsize=100)
    def abstract_method(self, token):
        pass

child_class = ChildClass()
print(child_class.abstract_method) # prints <functools._lru_cache_wrapper object at 0x12ad52820>
```

In pydantic V2 however
```python
class AbstractClass(BaseModel, ABC):

    @lru_cache(maxsize=100)
    @abstractmethod
    def abstract_method(self, token):
        pass


class ChildClass(AbstractClass):

    @lru_cache(maxsize=100)
    def abstract_method(self, token):
        pass

child_class = ChildClass()
print(child_class.abstract_method) # prints <bound method ChildClass.abstract_method of ChildClass()>
```

If your approach of using the `lru_cache` decorator works, there's an ongoing project to move to pydantic v2 and I'm wondering if need to do this after that is done. @masenf do you have any thoughts?

**Comment by AmanSal1** [comment]
@ElijahAhianyo I tried your solution, and yes, the initialization error is removed, but now there is another error. Maybe I am not writing the correct test case for the LRU caching
ERROR Message 
**FAILED test_state.py::test_lru_caching_memory - TypeError: StateManagerMemory.get_state() missing 1 required positional argument: 'token'**
